metrics.go

```diff
@@ -103,45 +109,57 @@ func (c *aggregatingCollector) Describe(descs chan<- *prometheus.Desc) {
 	}
 }
 
-func (c *aggregatingCollector) Collect(metrics chan<- prometheus.Metric) {
-	now := time.Now()
-	c.mutex.Lock()
-	var segsSent, segsRcvd uint64
-	var bytesSent, bytesRcvd uint64
-	for _, conn := range c.conns {
-		info, err := conn.getTCPInfo()
-		if err != nil {
-			if strings.Contains(err.Error(), "use of closed network connection") {
-				c.closedConn(conn)
+func (c *aggregatingCollector) cron() {
+	ticker := time.NewTicker(collectFrequency)
+	defer ticker.Stop()
+
+	for now := range ticker.C {
```

Comment:
Making a syscall every 10 seconds for every single connection seems like a non-starter, right? Most of these connections will be idle and we can have 10-100k connections on some machines.

This just seems like it's going to be a _major_ source of background CPU usage even when we're not using metrics.

Can we instead hook into the call to `Close` (we already have a `tracingConn`)?

Reply:
The exact same code was executed every time Prometheus scraped metrics, which by default happens every 15s. 

We _could_ hook into `Close`, but then we'd need special treatment for cases where the peer closed the connection. More importantly, we also need the syscall for collecting RTT and packet statistics.

Reply:
I wish there was a way to only execute this when Prometheus is actually running. Do we have a build time / runtime flag we can use?

Reply:
> The exact same code was executed every time Prometheus scraped metrics, which by default happens every 15s.

Only for nodes that actually care about metrics.

> We could hook into Close, but then we'd need special treatment for cases where the peer closed the connection. More importantly, we also need the syscall for collecting RTT and packet statistics.

We _always_ call close, even when the peer closes, to cleanup the connection (sockets, etc.).

> I wish there was a way to only execute this when Prometheus is actually running. Do we have a build time / runtime flag we can use?

We could wait for the first call to `Collect`, then stop when `Collect isn't called for a while. That's the best I can think of.